acme/challenge.go

```diff
@@ -344,6 +358,148 @@ func (hc *http01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo valida
 	return upd, nil
 }
 
+type tlsALPN01Challenge struct {
+	*baseChallenge
+}
+
+// newTLSALPN01Challenge returns a new acme tls-alpn-01 challenge.
+func newTLSALPN01Challenge(db nosql.DB, ops ChallengeOptions) (challenge, error) {
+	bc, err := newBaseChallenge(ops.AccountID, ops.AuthzID)
+	if err != nil {
+		return nil, err
+	}
+	bc.Type = "tls-alpn-01"
+	bc.Value = ops.Identifier.Value
+
+	hc := &tlsALPN01Challenge{bc}
+	if err := hc.save(db, nil); err != nil {
+		return nil, err
+	}
+	return hc, nil
+}
+
+func (tc *tlsALPN01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo validateOptions) (challenge, error) {
+	// If already valid or invalid then return without performing validation.
+	if tc.getStatus() == StatusValid || tc.getStatus() == StatusInvalid {
+		return tc, nil
+	}
+
+	config := &tls.Config{
+		NextProtos:         []string{"acme-tls/1"},
+		ServerName:         tc.Value,
+		InsecureSkipVerify: true, // we expect a self-signed challenge certificate
+	}
+
+	hostPort := net.JoinHostPort(tc.Value, "443")
+
+	conn, err := vo.tlsDial("tcp", hostPort, config)
+	if err != nil {
+		if err = tc.storeError(db,
+			ConnectionErr(errors.Wrapf(err, "error doing TLS dial for %s", hostPort))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+	defer conn.Close()
+
+	cs := conn.ConnectionState()
+	certs := cs.PeerCertificates
+
+	if len(certs) == 0 {
+		// note: it does not seem to be possible to trigger this path, as the Go TLS client will return a dial error
+		// when no certificate is served
+
+		if err = tc.storeError(db,
+			RejectedIdentifierErr(errors.Errorf("%s challenge for %s resulted in no certificates",
+				tc.Type, tc.Value))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+
+	if !cs.NegotiatedProtocolIsMutual || cs.NegotiatedProtocol != "acme-tls/1" {
+		if err = tc.storeError(db,
+			RejectedIdentifierErr(errors.Errorf("cannot negotiate ALPN acme-tls/1 protocol for "+
+				"tls-alpn-01 challenge"))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+
+	leafCert := certs[0]
+
+	if len(leafCert.DNSNames) != 1 || !strings.EqualFold(leafCert.DNSNames[0], tc.Value) {
+		if err = tc.storeError(db,
+			RejectedIdentifierErr(errors.Errorf("incorrect certificate for tls-alpn-01 challenge: "+
+				"leaf certificate must contain a single DNS name, %v", tc.Value))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+
+	idPeAcmeIdentifier := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 31}
+	idPeAcmeIdentifierV1Obsolete := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 30, 1}
```

Comment:
It makes me a little bit sad to see this obsolete OID propagate further. Is backwards compatibility with incorrect TLS-ALPN-01 challenge responses important for Smallstep? The larger ACME ecosystem might benefit from a nudge in the right direction by taking a firmer stance in new code and only expecting `idPeAcmeIdentifier`.

For reference, the Let's Encrypt metrics show Boulder sees the obsolete OID ~25-30% of the time. Not an insubstantial amount, but also not a majority.

Reply:
I don't have an opinion - if you are going to be the one merging this PR (hopefully) and prefer me to remove it I'm happy to oblige, just let me know. 

It would seem wise though to still check for this value and return a descriptive error if validation fails because of it, WDYT?

Reply:
Talked with @mmalone rather briefly, so apologies if we flip-flop, but we're cool with removing support in favor of a helpful error message, at least initially. If it's too frustrating for people, it's not hard at all to add back. Thanks @cpu for the critical eye.

Reply:
No problem, fixed it.

Reply:
So one interesting thing I just found out working on this is that the [golang.org/x/crypto/acme](https://github.com/golang/crypto/blob/master/acme/acme.go#L770) package is still generating certs with the obsolete identifier... I'm going to see if it can be updated :)

Reply:
@ibrt golang team welcomes contributions, you will need to create an issue first in golang/go, if it's not there yet, and then submit a PR, using GitHub, but by personal experience, I would recommend you to use directly Gerrit.
https://golang.org/doc/contribute.html

Reply:
They actually had a pending PR for that which has now been merged: https://github.com/golang/crypto/pull/91 :)